qiskit/quantum_info/operators/dihedral/dihedral.py

```diff
@@ -0,0 +1,463 @@
+            self.poly.set_term(list(j), (value + -2 * k) % 8)
+        for j in subsets_3:
+            value = self.poly.get_term(list(j))
+            self.poly.set_term(list(j), (value + 4 * k) % 8)
+
+    def _append_x(self, i):
+        """Apply X to this element.
+        Left multiply the element by X(i).
+        """
+        if not 0 <= i < self.num_qubits:
+            raise QiskitError("X qubit out of bounds.")
+        self.shift[i] = (self.shift[i] + 1) % 2
+
+    def __str__(self):
+        """Return formatted string representation."""
+        out = "phase polynomial = \n"
+        out += str(self.poly)
+        out += "\naffine function = \n"
+        out += " ("
+        for row in range(self.num_qubits):
+            wrote = False
+            for col in range(self.num_qubits):
+                if self.linear[row][col] != 0:
+                    if wrote:
+                        out += (" + x_" + str(col))
+                    else:
+                        out += ("x_" + str(col))
+                        wrote = True
+            if self.shift[row] != 0:
+                out += " + 1"
+            if row != self.num_qubits - 1:
+                out += ","
+        out += ")\n"
+        return out
+
+    def to_circuit(self):
+        """Return a QuantumCircuit implementing the CNOT-Dihedral element.
+
+        Return:
+            QuantumCircuit: a circuit implementation of the CNOTDihedral object.
+
+        References:
+            1. Shelly Garion and Andrew W. Cross, *Synthesis of CNOT-Dihedral circuits
+               with optimal number of two qubit gates*, `Quantum 4(369), 2020
+               <https://quantum-journal.org/papers/q-2020-12-07-369/>`_
+            2. Andrew W. Cross, Easwar Magesan, Lev S. Bishop, John A. Smolin and Jay M. Gambetta,
+               *Scalable randomised benchmarking of non-Clifford gates*,
+               npj Quantum Inf 2, 16012 (2016).
+        """
+        return decompose_cnotdihedral(self)
+
+    def to_instruction(self):
+        """Return a Gate instruction implementing the CNOTDihedral object."""
+        return self.to_circuit().to_gate()
+
+    def _from_circuit(self, circuit):
+        """Initialize from a QuantumCircuit or Instruction.
+
+        Args:
+            circuit (QuantumCircuit or ~qiskit.circuit.Instruction):
+                instruction to initialize.
+        Returns:
+            CNOTDihedral: the CNOTDihedral object for the circuit.
+        Raises:
+            QiskitError: if the input instruction is not CNOTDihedral or contains
+                         classical register instruction.
+        """
+        if not isinstance(circuit, (QuantumCircuit, Instruction)):
+            raise QiskitError("Input must be a QuantumCircuit or Instruction")
+
+        # Initialize an identity CNOTDihedral object
+        elem = CNOTDihedral(num_qubits=self._num_qubits)
+        _append_circuit(elem, circuit)
+        return elem
+
+    def to_matrix(self):
```

Comment:
It would be good to also add an `__array__` method to so it can be used with numpy functions. You can copy implementation from Pauli class
```python
def __array__(self, dtype=None):
        if dtype:
            return np.asarray(self.to_matrix(), dtype=dtype)
        return self.to_matrix()
```